state/cluster/badger/mutator.go

```diff
@@ -36,202 +37,286 @@ func NewMutableState(state *State, tracer module.Tracer, headers storage.Headers
+	parentID := header.ParentID
+	for parentID != extendCtx.finalizedClusterBlock.ID() {
 
-		// extending block must have correct parent view
-		if header.ParentView != parent.View {
-			return state.NewInvalidExtensionErrorf("candidate build with inconsistent parent view (candidate: %d, parent %d)",
-				header.ParentView, parent.View)
+		// get the parent of current block
+		ancestor, err := m.headers.ByBlockID(parentID)
+		if err != nil {
+			return fmt.Errorf("could not get parent (%x): %w", block.Header.ParentID, err)
 		}
 
-		// the extending block must increase height by 1 from parent
-		if header.Height != parent.Height+1 {
-			return state.NewInvalidExtensionErrorf("extending block height (%d) must be parent height + 1 (%d)",
-				block.Header.Height, parent.Height)
+		// if its height is below current boundary, the block does not connect
+		// to the finalized protocol state and would break database consistency
+		if ancestor.Height < extendCtx.finalizedClusterBlock.Height {
+			return state.NewOutdatedExtensionErrorf("block doesn't connect to finalized state. ancestor.Height (%d), final.Height (%d)",
+				ancestor.Height, extendCtx.finalizedClusterBlock.Height)
 		}
 
-		// ensure that the extending block connects to the finalized state, we
-		// do this by tracing back until we see a parent block that is the
-		// latest finalized block, or reach height below the finalized boundary
-
-		setupSpan.End()
-		checkAnsSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendCheckAncestry)
+		parentID = ancestor.ParentID
+	}
+	checkAncestrySpan.End()
 
-		// start with the extending block's parent
-		parentID := header.ParentID
-		for parentID != finalizedClusterBlock.ID() {
+	checkRefBlockSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendCheckReferenceBlock)
+	err = m.checkPayloadReferenceBlock(extendCtx)
+	if err != nil {
+		return fmt.Errorf("invalid reference block: %w", err)
+	}
+	checkRefBlockSpan.End()
 
-			// get the parent of current block
-			ancestor, err := m.headers.ByBlockID(parentID)
-			if err != nil {
-				return fmt.Errorf("could not get parent (%x): %w", block.Header.ParentID, err)
-			}
+	checkTxsSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendCheckTransactionsValid)
+	err = m.checkPayloadTransactions(extendCtx)
+	if err != nil {
+		return fmt.Errorf("invalid payload transactions: %w", err)
+	}
+	checkTxsSpan.End()
 
-			// if its height is below current boundary, the block does not connect
-			// to the finalized protocol state and would break database consistency
-			if ancestor.Height < finalizedClusterBlock.Height {
-				return state.NewOutdatedExtensionErrorf("block doesn't connect to finalized state. ancestor.Height (%d), final.Height (%d)",
-					ancestor.Height, finalizedClusterBlock.Height)
-			}
+	insertDbSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendDBInsert)
```

Comment:
I am curious why we use different span names? Isn't this essentially the same operation. We already have context.

Reply:
I'm not following what you mean. What is this the same operation as?

Reply:
I'm going to merge this PR, but if there is any change you'd like to make here related to this comment, please let me know and I can incorporate it into https://github.com/onflow/flow-go/pull/4296

Reply:
nvm. not understanding my own comment either (anymore) 😅